perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/VosManagerBlImpl.java

```diff
@@ -902,14 +928,54 @@ public boolean usesEmbeddedGroupRegistrations(PerunSession sess, Vo vo) {
 
 	@Override
 	public void addMemberVo(PerunSession sess, Vo vo, Vo memberVo) throws RelationExistsException {
-		// todo - add necessary logic for adding new member vo
 		vosManagerImpl.addMemberVo(sess, vo, memberVo);
+
+		List<Member> memberVoMembers = perunBl.getMembersManagerBl().getMembers(sess, memberVo);
+
+		for (Member member : memberVoMembers) {
+			try {
+
+				try {
+					// if user is member in both vos, update memberOrganizations list attribute
+					Member existingMember = perunBl.getMembersManagerBl().getMemberByUserId(sess, vo, member.getUserId());
+					Attribute attribute = perunBl.getAttributesManagerBl().getAttribute(sess, existingMember, A_MEMBER_DEF_MEMBER_ORGANIZATIONS);
+					ArrayList<String> currentValue = attribute.valueAsList();
+					currentValue.add(memberVo.getShortName());
+					attribute.setValue(currentValue);
+					perunBl.getAttributesManagerBl().setAttribute(sess, existingMember, attribute);
+
+				} catch (MemberNotExistsException e) {
+					// if user is member only in member vo, create member in parent vo and set up memberOrganizations list attribute
+					Member newMember = perunBl.getMembersManagerBl().createMember(sess, vo, perunBl.getUsersManagerBl().getUserByMember(sess, member));
+					perunBl.getMembersManagerBl().validateMemberAsync(sess, member);
```

Comment:
I am not sure if we can validate the members asynchronously. The async method waits at first for 5 sec and then starts validating. It waits 5 seconds because the transaction, which inserts the member to DB, might not have been committed yet. In our case, I am not sure if the transaction will be committed within these 5 seconds if there are thousands of members to be created.

Reply:
We will test this and adjust the code in separate PR.